File: apps/mobile/lib/core/utils/operating_hours.dart
// Operating hours model — mirrors the JSONB structure in store_settings.operating_hours.
//
// Example:
// { "mon": {"open": "11:00", "close": "22:00"}, "tue": {...}, ... }

typedef DayHours = ({String open, String close});
typedef OperatingHours = Map<String, DayHours>;

const _dayKeys = ['mon', 'tue', 'wed', 'thu', 'fri', 'sat', 'sun'];

/// Parses operating hours from a Map (Supabase JSONB).
OperatingHours parseOperatingHours(Map<String, dynamic> json) {
  return json.map((day, value) {
    final map = value as Map<String, dynamic>;
    return MapEntry(day, (open: map['open'] as String, close: map['close'] as String));
  });
}

/// Returns the day key for the given [dateTime] (e.g. "mon", "tue").
String _dayKey(DateTime dateTime) {
  return _dayKeys[dateTime.weekday - 1]; // weekday is 1=Mon..7=Sun
}

/// Checks if the store is currently open based on [operatingHours].
bool isStoreOpen(OperatingHours operatingHours, [DateTime? now]) {
  now ??= DateTime.now();
  final dayKey = _dayKey(now);
  final hours = operatingHours[dayKey];
  if (hours == null) return false;

  final currentMinutes = now.hour * 60 + now.minute;
  final openMinutes = _parseTimeToMinutes(hours.open);
  final closeMinutes = _parseTimeToMinutes(hours.close);

  return currentMinutes >= openMinutes && currentMinutes < closeMinutes;
}

/// Returns a human-readable string for when the store next opens.
/// Returns null if the store is currently open.
String? formatNextOpenTime(OperatingHours operatingHours, [DateTime? now]) {
  now ??= DateTime.now();
  if (isStoreOpen(operatingHours, now)) return null;

  final todayKey = _dayKey(now);
  final todayHours = operatingHours[todayKey];
  if (todayHours != null) {
    final openMinutes = _parseTimeToMinutes(todayHours.open);
    final currentMinutes = now.hour * 60 + now.minute;
    if (currentMinutes < openMinutes) {
      return 'Opens today at ${todayHours.open}';
    }
  }

  // Check upcoming days
  for (int i = 1; i <= 7; i++) {
    final futureDate = now.add(Duration(days: i));
    final futureKey = _dayKey(futureDate);
    final futureHours = operatingHours[futureKey];
    if (futureHours != null) {
      final dayLabel = i == 1 ? 'tomorrow' : _dayFullName(futureKey);
      return 'Opens $dayLabel at ${futureHours.open}';
    }
  }

  return 'Currently closed';
}

int _parseTimeToMinutes(String time) {
  final parts = time.split(':');
  return int.parse(parts[0]) * 60 + int.parse(parts[1]);
}

String _dayFullName(String key) {
  return const {
    'mon': 'Monday',
    'tue': 'Tuesday',
    'wed': 'Wednesday',
    'thu': 'Thursday',
    'fri': 'Friday',
    'sat': 'Saturday',
    'sun': 'Sunday',
  }[key]!;
}

File: apps/web/src/lib/utils/operating-hours.ts
export interface DayHours {
  open: string;
  close: string;
}

export interface OperatingHours {
  mon: DayHours;
  tue: DayHours;
  wed: DayHours;
  thu: DayHours;
  fri: DayHours;
  sat: DayHours;
  sun: DayHours;
}

export interface StoreStatus {
  isOpen: boolean;
  nextOpen?: {
    day: string;
    time: string;
    date: Date;
  };
}

function getDayKey(date: Date): keyof OperatingHours {
  const days: (keyof OperatingHours)[] = ['sun', 'mon', 'tue', 'wed', 'thu', 'fri', 'sat'];
  return days[date.getDay()];
}

function parseTime(timeStr: string): { hours: number; minutes: number } {
  const [hours, minutes] = timeStr.split(':').map(Number);
  return { hours, minutes };
}

function timeToMinutes(timeStr: string): number {
  const { hours, minutes } = parseTime(timeStr);
  return hours * 60 + minutes;
}

function formatDayName(day: keyof OperatingHours): string {
  const dayNames: Record<keyof OperatingHours, string> = {
    mon: 'Monday',
    tue: 'Tuesday',
    wed: 'Wednesday',
    thu: 'Thursday',
    fri: 'Friday',
    sat: 'Saturday',
    sun: 'Sunday',
  };
  return dayNames[day];
}

export function isStoreOpen(
  operatingHours: OperatingHours | null,
  now: Date = new Date()
): StoreStatus {
  if (!operatingHours) {
    return { isOpen: false };
  }

  const currentDay = getDayKey(now);
  const currentMinutes = now.getHours() * 60 + now.getMinutes();
  const todayHours = operatingHours[currentDay];

  if (!todayHours) {
    return { isOpen: false };
  }

  const openMinutes = timeToMinutes(todayHours.open);
  const closeMinutes = timeToMinutes(todayHours.close);

  if (currentMinutes >= openMinutes && currentMinutes < closeMinutes) {
    return { isOpen: true };
  }

  const days: (keyof OperatingHours)[] = ['mon', 'tue', 'wed', 'thu', 'fri', 'sat', 'sun'];
  const currentDayIndex = days.indexOf(currentDay);

  for (let i = 1; i <= 7; i++) {
    const nextDayIndex = (currentDayIndex + i) % 7;
    const nextDay = days[nextDayIndex];
    const nextDayHours = operatingHours[nextDay];

    if (nextDayHours && nextDayHours.open !== nextDayHours.close) {
      const nextDate = new Date(now);
      nextDate.setDate(now.getDate() + i);

      return {
        isOpen: false,
        nextOpen: {
          day: formatDayName(nextDay),
          time: nextDayHours.open,
          date: nextDate,
        },
      };
    }
  }

  return { isOpen: false };
}

export function formatNextOpenTime(nextOpen: StoreStatus['nextOpen']): string {
  if (!nextOpen) return '';

  const today = new Date();
  const tomorrow = new Date(today);
  tomorrow.setDate(today.getDate() + 1);

  let dayPrefix = nextOpen.day;
  if (nextOpen.date.toDateString() === today.toDateString()) {
    dayPrefix = 'Today';
  } else if (nextOpen.date.toDateString() === tomorrow.toDateString()) {
    dayPrefix = 'Tomorrow';
  }

  return `${dayPrefix} at ${nextOpen.time}`;
}
